File: app/src/main/java/com/example/instagram/login/LoginActivity.java
package com.example.instagram.login;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.example.instagram.main.MainActivity;
import com.example.instagram.R;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.google.android.material.textview.MaterialTextView;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseAuthInvalidCredentialsException;

public class LoginActivity extends AppCompatActivity {
    private FirebaseAuth firebaseAuth = FirebaseAuth.getInstance();

    MaterialTextView materialTextView;
    TextInputLayout email_text_layout, pass_text_layout;
    TextInputEditText pass_text, mail_text;

    MaterialButton login_btn;
    MaterialButton create_acc_btn;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        //EdgeToEdge.enable(this);
        setContentView(R.layout.activity_login);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });
        
        // login with validation
        
        login_btn = findViewById(R.id.login_btn);
        login_btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mail_text.getText().toString().trim().isEmpty()){
                    email_text_layout.setError("Mail cannot be empty");
                } else if (pass_text.getText().toString().trim().isEmpty()){
                    pass_text_layout.setError("Password required");
                }else {
                    signIn(mail_text.getText().toString().trim(), pass_text.getText().toString().trim());
                }

            }
        });


        // button handling
        login_btn = findViewById(R.id.login_btn);
        create_acc_btn = findViewById(R.id.create_acc_btn);

        create_acc_btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(LoginActivity.this, CreateAccActivity.class);
                startActivity(intent);
            }
        });

        // set position start

        mail_text = findViewById(R.id.mail_text);
        pass_text = findViewById(R.id.pass_text);

        mail_text.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if(!hasFocus){
                    mail_text.scrollTo(0, 0);
                }
            }
        });

        pass_text.setOnFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {
                if(!hasFocus){
                    pass_text.scrollTo(0, 0);
                }
            }
        });


        // login validation
        email_text_layout = findViewById(R.id.email_text_layout);
        pass_text_layout = findViewById(R.id.pass_text_layout);
        pass_text = findViewById(R.id.pass_text);
        mail_text = findViewById(R.id.mail_text);

        mail_text.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                String text = s.toString().trim();
                mailValidation(text);
            }
        });
        

        pass_text.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                String password = pass_text.getText().toString().trim();
                passValidation(password);
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

    }

    private void signIn(String email, String password) {
        email_text_layout.setError(null);
        pass_text_layout.setError(null);
        firebaseAuth.signInWithEmailAndPassword(email, password).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
            @Override
            public void onComplete(@NonNull Task<AuthResult> task) {
                if (task.isSuccessful()){
                    Intent intent = new Intent(LoginActivity.this, MainActivity.class);
                    startActivity(intent);
                    Toast.makeText(LoginActivity.this, "Logged In", Toast.LENGTH_SHORT).show();
                    finish();
                } else if (task.getException() instanceof FirebaseAuthInvalidCredentialsException) {
                    String errorCode = ((FirebaseAuthInvalidCredentialsException) task.getException()).getErrorCode();
                    if (!errorCode.isEmpty()){
                        Toast.makeText(LoginActivity.this, "Invalid Credentials", Toast.LENGTH_SHORT).show();
                    }
                }else {
                    Toast.makeText(LoginActivity.this, "Login Failed", Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    private boolean mailValidation(String text){
        String regex = "^[a-zA-Z0-9_%.+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,6}$";
        if (text.isEmpty()){
            email_text_layout.setError("Email cannot be empty");
            return false;
        } else if (!text.matches(regex)) {
            email_text_layout.setError("Invalid email address");
            return false;
        }else {
            email_text_layout.setError(null);
            return true;
        }
    }
    
    private boolean passValidation(String text){
        String regex = "^[a-zA-Z0-9@#$&_.]+$";
        if (text.isEmpty()) {
            pass_text_layout.setError("Password required");
            return false;
        } else if (!text.matches(regex)){
            pass_text_layout.setError("Enter valid password");
            return false;
        }else {
            pass_text_layout.setError(null);
            return true;
        }
    }
}

File: app/src/main/java/com/example/ktinsta/login/LoginActivity.kt
package com.example.ktinsta.login

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.example.ktinsta.main.MainActivity
import com.example.instagram.R
import com.google.android.material.button.MaterialButton
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuthInvalidCredentialsException

class LoginActivity : AppCompatActivity() {
    private val firebaseAuth: FirebaseAuth = FirebaseAuth.getInstance()

    private lateinit var emailTextLayout: TextInputLayout
    private lateinit var passTextLayout: TextInputLayout
    private lateinit var passText: TextInputEditText
    private lateinit var mailText: TextInputEditText
    private lateinit var loginBtn: MaterialButton
    private lateinit var createAccBtn: MaterialButton

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        // Initialize views
        emailTextLayout = findViewById(R.id.email_text_layout)
        passTextLayout = findViewById(R.id.pass_text_layout)
        passText = findViewById(R.id.pass_text)
        mailText = findViewById(R.id.mail_text)
        loginBtn = findViewById(R.id.login_btn)
        createAccBtn = findViewById(R.id.create_acc_btn)

        // Login button click listener
        loginBtn.setOnClickListener {
            val email = mailText.text.toString().trim()
            val password = passText.text.toString().trim()

            if (!mailValidation(email) || !passValidation(password)) {
                return@setOnClickListener
            }
            signIn(email, password)
        }

        // Create account button click listener
        createAccBtn.setOnClickListener {
            startActivity(Intent(this, CreateAccActivity::class.java))
        }

        // Prevent scrolling when losing focus
        val focusChangeListener = View.OnFocusChangeListener { v, hasFocus ->
            if (!hasFocus) v.scrollTo(0, 0)
        }
        mailText.onFocusChangeListener = focusChangeListener
        passText.onFocusChangeListener = focusChangeListener

        // Add text watchers for validation
        mailText.addTextChangedListener(validationTextWatcher { text ->
            mailValidation(text)
        })
        passText.addTextChangedListener(validationTextWatcher { text ->
            passValidation(text)
        })
    }

    private fun signIn(email: String, password: String) {
        emailTextLayout.error = null
        passTextLayout.error = null

        firebaseAuth.signInWithEmailAndPassword(email, password).addOnCompleteListener { task ->
            if (task.isSuccessful) {
                startActivity(Intent(this, MainActivity::class.java))
                Toast.makeText(this, "Logged In", Toast.LENGTH_SHORT).show()
                finish()
            } else {
                val errorMessage = when (task.exception) {
                    is FirebaseAuthInvalidCredentialsException -> "Invalid Credentials"
                    else -> "Login Failed"
                }
                Toast.makeText(this, errorMessage, Toast.LENGTH_SHORT).show()
            }
        }
    }

    private fun mailValidation(text: String): Boolean {
        val regex = "^[a-zA-Z0-9_%.+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,6}$".toRegex()
        return when {
            text.isEmpty() -> {
                emailTextLayout.error = "Email cannot be empty"
                false
            }
            !text.matches(regex) -> {
                emailTextLayout.error = "Invalid email address"
                false
            }
            else -> {
                emailTextLayout.error = null
                true
            }
        }
    }

    private fun passValidation(text: String): Boolean {
        val regex = "^[a-zA-Z0-9@#$&_.]+$".toRegex()
        return when {
            text.isEmpty() -> {
                passTextLayout.error = "Password required"
                false
            }
            !text.matches(regex) -> {
                passTextLayout.error = "Enter valid password"
                false
            }
            else -> {
                passTextLayout.error = null
                true
            }
        }
    }

    private fun validationTextWatcher(validate: (String) -> Boolean): TextWatcher {
        return object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                validate(s?.toString()?.trim().orEmpty())
            }

            override fun afterTextChanged(s: Editable?) {}
        }
    }
}
